includes/promoted-jobs/class-wp-job-manager-promoted-jobs-notifications.php

```diff
@@ -0,0 +1,335 @@
+			$post_data[ $field ] = isset( $_POST[ "_$field" ] ) ? wp_unslash( $_POST[ "_$field" ] ) : '';
+		}
+		return $post_data;
+	}
+
+	/**
+	 * Determine if any job field has changed.
+	 *
+	 * @param array $current_values Current values of the job.
+	 * @param array $new_values New values of the job.
+	 */
+	private function post_has_changed( $current_values, $new_values ) {
+		foreach ( $current_values as $key => $value ) {
+			if ( $value !== $new_values[ $key ] ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check if a job is promoted.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return bool
+	 */
+	private function is_promoted_job( $post_id ) {
+		return get_post_meta( $post_id, '_promoted', true );
+	}
+
+	/**
+	 * Check if we should notify wpjobmanager.com of a change.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return bool
+	 */
+	public function should_notify( $post_id ) {
+		if ( 'job_listing' !== get_post_type( $post_id ) ) {
+			return false;
+		}
+		if ( ! $this->is_promoted_job( $post_id ) ) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing is trashed.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return void
+	 */
+	public function promoted_job_trashed( $post_id ) {
+		if ( ! $this->should_notify( $post_id ) ) {
+			return;
+		}
+		$this->send_notification();
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing is updated.
+	 *
+	 * @param int   $post_id Post ID.
+	 * @param array $values Values.
+	 * @return void
+	 */
+	public function promoted_job_updated( $post_id, $values ) {
+		if ( ! $this->should_notify( $post_id ) ) {
+			return;
+		}
+
+		if ( $this->post_has_changed( $this->get_watched_fields_meta( $post_id ), $this->map_to_watched_fields( $values ) ) ) {
+			$this->send_notification();
+		}
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing is updated on admin.
+	 * Values are available in the $_POST global, sanitized and validated.
+	 *
+	 * @param int     $post_id Post ID.
+	 * @param WP_Post $post Post object.
+	 * @return void
+	 */
+	public function promoted_job_updated_admin( $post_id, $post ) {
+		if ( ! $this->should_notify( $post_id ) ) {
+			return;
+		}
+
+		if ( $this->post_has_changed( $this->get_watched_fields_meta( $post_id ), $this->get_post_data( $post ) ) ) {
+			$this->send_notification();
+		}
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing changes.
+	 *
+	 * @return void
+	 */
+	private function send_notification() {
+		$response = wp_remote_post( $this->get_notification_url() );
+		$this->maybe_schedule_rerun( $response );
+	}
+
+	/**
+	 * Schedule a cron job to run the notification task if API call failed, unschedule if it succeeds.
+	 *
+	 * @param WP_Error|array $response Response from wp_remote_post.
+	 * @return void
+	 */
+	public function maybe_schedule_rerun( $response ) {
```

Comment:
Do you think the retry could be simplified a little using an approach similar to [this one](https://github.com/Automattic/wpjobmanager.com/pull/444/files#diff-205ccc80cb5c8b42278efe5172eee92cf972cd0476825c0cb6b013c2a49f13c3R112-R119) (basically scheduling a single job after each fail instead a cron job and control the number of retries through its arguments)?

Reply:
Yeah, much simpler in baf019ff677d12aedc7a154efdce9ed17f194328

Reply:
Happy that it could be simplified! =)

While reviewing your PR, I just realized a case that we need to handle and I also fixed it in my PR.

If it fails and already has a retry scheduled, we don't need to schedule another one. And if it works, we can cancel existing retries.

For reference, it's what I did in my PR: https://github.com/Automattic/wpjobmanager.com/pull/444/commits/fa2af0ace56b89b8d006218457f2edb7d06a37c4, but I think you can make it simpler here, not needing to check the arguments.

In my PR it was more complicated because I couldn't clear all the hooks because we have arguments for different sites. In your PR, I think you can handle it more simply, checking and canceling all hooks with the retry name `wp_next_scheduled( self::RETRY_JOB_NAME )` / `wp_clear_scheduled_hook( self::RETRY_JOB_NAME )`. WDYT?

Reply:
Added in a6b619b2b8718671e7c4b2dc133f96f399c47dd2

Note however that I'm using `wp_unschedule_hook` to cancel. 
Both `wp_clear_scheduled_hook` and `wp_next_scheduled` actually require the exact arguments(ie we'll need to loop through all retry values as in your PR). 
With `wp_unschedule_hook`, it can all be cleared on one line.

Reply:
Ah! TIL! I didn't know it was a different function when using it without arguments. Do you think we also have a similar one for the `wp_next_scheduled` to avoid the other loop? I tried to Google quickly, but I didn't find anything.

Reply:
Yeah I checked too check too but didn't find any.